usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2298 @@
+        *version = TOKVERSION_312;
+        goto done;
+    }
+
+    /* Read old object */
+    *version = TOKVERSION_00;
+    ret = read_object_00(fp, name, size, obj, obj_len, obj_priv);
+    if (ret != 0) {
+        TRACE_ERROR("Cannot read old object %s.\n", name);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    ret = CKR_OK;
+
+done:
+
+    if (fp) {
+        fclose(fp);
+        fp = NULL;
+    }
+
+    return ret;
+}
+
+/**
+ * Migrate the token objects from old to new format. Some of the token objects
+ * may be in old and some may be in new format if a previous migration run
+ * was interrupted.
+ */
+CK_RV migrate_token_objects(const char *data_store, const CK_BYTE *masterkey_old,
+                            const CK_BYTE *masterkey_new,
+                            const CK_BYTE *so_wrap_key,
+                            const CK_BYTE *user_wrap_key)
+{
+    FILE *fp = NULL;
+    unsigned char *obj = NULL;
+    unsigned int obj_len;
+    char tmp[PATH_MAX], iname[PATH_MAX];
+    CK_BBOOL priv;
+    CK_ULONG version;
+    int count = 0, scount = 0;
+    CK_RV ret;
+
+    /* Check parms */
+    if (!data_store || !masterkey_old || !masterkey_new || !so_wrap_key
+        || !user_wrap_key) {
+        TRACE_ERROR("Invalid parms.\n");
+        ret = CKR_ARGUMENTS_BAD;
+        goto done;
+    }
+
+    /* Open index file OBJ.IDX */
+    snprintf(iname, sizeof(iname), "%s/TOK_OBJ/OBJ.IDX", data_store);
+    fp = fopen((char *)iname, "r");
+    if (!fp) {
+        TRACE_ERROR("Cannot open %s\n",iname);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Migrate items from OBJ.IDX */
+    while (fgets(tmp, 50, fp)) {
+        tmp[strlen(tmp) - 1] = 0;
+        ret = read_object(data_store, tmp, &obj, &obj_len, &version, &priv);
+        if (ret == 0 && version == TOKVERSION_00) {
+            if (priv) {
+                ret = migrate_private_token_object(data_store, tmp,
+                                   obj, obj_len, masterkey_old, masterkey_new);
+                if (ret != 0) {
+                    TRACE_ERROR("Cannot migrate private object %s, continuing ... \n", tmp);
+                } else
+                    scount++;
+            } else {
+                ret = migrate_public_token_object(data_store, tmp,
+                                                  obj, obj_len);
+                if (ret != 0) {
+                    TRACE_ERROR("Cannot migrate public object %s, continuing ... \n", tmp);
+                } else
+                    scount++;
+            }
+        }
+
+        if (obj) {
+            free(obj);
+            obj = NULL;
+        }
+        count++;
+    }
+
+    /* Close OBJ.IDX */
+    fclose(fp);
+
+    ret = CKR_OK;
+
+    TRACE_NONE("Migrated %d object(s) out of %d object(s).\n", scount, count);
+
+done:
+
+    return ret;
+}
+
+/**
+ * loads the new aes256 masterkey
+ */
+CK_RV load_masterkey_312(const char *data_store, const char * mkfile,
```

Comment:
and probably same for these masterkey functions

Reply:
so essentially the migrate tool would only have the migrate_xxx functions locally, anything related to reading/making/decrypting would be shared code.

Reply:
yep.